crates/scheduler/src/batch_scheduler.rs

```diff
@@ -0,0 +1,448 @@
+use std::sync::Arc;
+
+use hypha_messages::progress::{self, Metrics};
+use hypha_network::request_response::{RequestResponseError, RequestResponseInterfaceExt};
+use libp2p::PeerId;
+use thiserror::Error;
+use tokio::{
+    sync::{
+        Mutex, Notify,
+        mpsc::{self, Sender, error::SendError},
+    },
+    task::JoinHandle,
+};
+use uuid::Uuid;
+
+use crate::{
+    network::Network,
+    simulation::SimulatorFn,
+    statistics::RuntimeStatistic,
+    tracker::{ProgressTracker, State, TrackerError},
+};
+
+#[derive(Debug, Error)]
+pub enum BatchSchedulerError {
+    #[error("Disconnected")]
+    Disconnected,
+    #[error("Unregistering worker")]
+    Unregister,
+    #[error("Error during scheduling {0}")]
+    Scheduling(String),
+    #[error("Network error")]
+    NetworkError(#[from] RequestResponseError),
+    #[error("Send Metrics Error")]
+    SendMetricsError(#[from] SendError<(PeerId, Metrics)>),
+    #[error("Tracker error")]
+    TrackerError(#[from] TrackerError),
+}
+
+/// The Batch Scheduler uses internally the following states
+/// for scheduling
+/// flowchart TD
+/// A[Training] --> |Worker: Status| B{Project}
+/// B --> |Not done| A
+/// B -->|Done| C[UpdateScheduled]
+/// C -->|Worker: Update| E[Updating]
+/// E -->|PS: Updated| F[Update Received]
+/// F --> A
+async fn handle_request<T>(
+    tx: Sender<(PeerId, Metrics)>,
+    tracker: Arc<Mutex<ProgressTracker<T>>>,
+    request: (PeerId, progress::Request),
+    simulator: SimulatorFn,
+    notifier: Arc<Notify>,
+) -> Result<progress::Response, BatchSchedulerError>
+where
+    T: RuntimeStatistic + 'static,
+{
+    let (
+        peer_id,
+        progress::Request {
+            progress: update, ..
+        },
+    ) = request;
+    tracing::info!(
+        %peer_id,
+        ?update,
+        "Received update",
+    );
+    match update {
+        progress::Progress::Metrics(metrics) => {
+            tx.send((peer_id, metrics))
+                .await
+                .map_err(BatchSchedulerError::from)?;
+            Ok(progress::Response::Ok {})
+        }
+        progress::Progress::Status(progress::Status { batch_size }) => {
+            let mut tracker = tracker.lock().await;
+            tracker.update(&peer_id, batch_size)?;
+            match tracker.worker_tracker.worker_state(&peer_id) {
+                State::Training => {
+                    let time_cap = 10000u64;
+                    let update_cap = 3u64;
```

Comment:
🤔 Feels like we should make these configurable or derive them from the scheduler recipe instead of hard-coding them - but probably fine for v0.